### PR TITLE
dev/database: Start DBSource.java & DBMapper.java for Storage and Retrieval of Customer Details

### PR DESCRIPTION
**Pull request by sgoudham**

### CODE REVIEW

File: src/main/java/org/somerville/swag/data/entity/Customer.java [hunk 1]
```diff
@@ -86,83 +79,29 @@ public void purchaseItems() {
...
-    }
-
-    public String getPhoneNumber() {
-        return phoneNumber;
-    }
-
-    public void setPhoneNumber(String phoneNumber) {
-        this.phoneNumber = phoneNumber;
-    }
+    public void setCurrentOrder(Order currentOrder) { this.currentOrder = currentOrder; }
```
Comment (sgoudham) [inline src/main/java/org/somerville/swag/data/entity/Customer.java:82]: Really liking the condensed version for getters/setters as it reduces lines of code!

File: src/main/java/org/somerville/swag/data/service/LoggingServiceImpl.java [hunk 2]
```diff
@@ -83,7 +83,31 @@ public void logDatabaseSelectSuccess(String selectQuery) {
 
     @Override
     public void logDatabaseSelectFailure(String selectQuery, String failureMessage) {
-        String logMessage  = events.getDatabaseReadFailure() + ": " + failureMessage + " - Statement: " + selectQuery;
+        String logMessage = events.getDatabaseReadFailure() + ": " + failureMessage + " - Statement: " + selectQuery;
```
Comment (sgoudham) [inline src/main/java/org/somerville/swag/data/service/LoggingServiceImpl.java:86]: Great job ensuring consistency across all logging events!

File: src/main/java/org/somerville/swag/data/entity/Customer.java [hunk 3]
```diff
@@ -86,83 +79,29 @@ public void purchaseItems() {
...
+    public int getCustomerId() { return customerId; }
+    public String getForename() { return forename; }
+    public String getSurname() { return surname; }
+    public String getEmail() { return email; }
+    public String getPassword() { return password; }
+    public String getAddressLine1() { return addressLine1; }
+    public String getAddressLine2() { return addressLine2; }
+    public String getCity() { return city; }
+    public String getPostcode() { return postcode; }
+    public String getPhoneNumber() { return phoneNumber; }
```
Comment (2013884f) [inline src/main/java/org/somerville/swag/data/entity/Customer.java:105]: This looks much cleaner.

**Review by 2013884f — APPROVED** [review]
The code for building the various log messages reads really well. 
The mocks for the product and customer mapping has been done very well.

File: src/main/java/org/somerville/swag/data/entity/Product.java [hunk 4]
```diff
@@ -10,69 +11,59 @@
...
+
+    public int getProductId() { return productId; }
+    public String getName() { return name; }
+    public String getDescription() { return description; }
+    public BigDecimal getPrice() { return price; }
+    public int getStockLevel() { return stockLevel; }
+    public String getImagePath() { return imagePath; }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (AlanRobinson1337) [inline src/main/java/org/somerville/swag/data/entity/Product.java:46]: Can we refactor variable "o" for increased readability?

File: src/main/java/org/somerville/swag/data/entity/Customer.java [hunk 5]
```diff
@@ -86,83 +79,29 @@ public void purchaseItems() {
...
+    public void setEmail(String email) { this.email = email; }
+    public void setPassword(String password) { this.password = password; }
+    public void setAddressLine1(String addressLine1) { this.addressLine1 = addressLine1; }
+    public void setAddressLine2(String addressLine2) { this.addressLine2 = addressLine2; }
+    public void setCity(String city) { this.city = city; }
+    public void setPostcode(String postcode) { this.postcode = postcode; }
+    public void setPhoneNumber(String phoneNumber) { this.phoneNumber = phoneNumber; }
+
+    public CustomerState getCustomerState() { return customerState; }
+    public Order getCurrentOrder() { return currentOrder; }
```
Comment (RobertRanson) [inline src/main/java/org/somerville/swag/data/entity/Customer.java:95]: This is much more readable, good change!

File: src/main/java/org/somerville/swag/data/entity/state/Guest.java [hunk 6]
```diff
@@ -45,7 +63,7 @@ public void removeProductFromBasket() {
 
     @Override
     public void purchaseItems() {
-
+        // TODO: Display error that you need to be logged in to checkout
```
Comment (RobertRanson) [inline src/main/java/org/somerville/swag/data/entity/state/Guest.java:66]: These TODO are really useful, it is clear what needs to be worked on next

File: src/main/java/org/somerville/swag/data/source/DBMapper.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    private LoggingService loggingService = LoggingServiceImpl.getInstance();
+
+    public DBMapper() { }
+
+    public void mapToCustomer(ResultSet customerData, Customer customer) {
+        try {
+            if (!customerData.isBeforeFirst()) {
+                loggingService.logDatabaseCustomerNotFound();
+            } else {
+                while (customerData.next()) {
```
Comment (AlanRobinson1337) [inline src/main/java/org/somerville/swag/data/source/DBMapper.java:23]: Look to remove while loop as part of future refactoring - unsure if resultSets throw errors without

**Review by AlanRobinson1337 — APPROVED** [review]
Much cleaner code for getters and setters across pull request
Great readability overall.
Additional class files look great

File: src/main/java/org/somerville/swag/data/entity/state/LoggedIn.java [hunk 8]
```diff
@@ -7,15 +7,15 @@
 
 public class LoggedIn implements CustomerState {
 
-    Customer customer;
+    private Customer customer;
```
Comment (RobertRanson) [inline src/main/java/org/somerville/swag/data/entity/state/LoggedIn.java:10]: This change helps our project follow the principles of encapsulation. nice!

**Review by RobertRanson — APPROVED** [review]

**Review by hzl616 — APPROVED** [review]
Easy to read, efficient use of variable names

File: src/main/java/org/somerville/swag/data/entity/Product.java (hunk 4 above)
Comment (sgoudham) [reply src/main/java/org/somerville/swag/data/entity/Product.java:46]: Good idea! Will definitely refactor it to be more readable

File: src/main/java/org/somerville/swag/data/source/DBMapper.java (hunk 7 above)
Comment (sgoudham) [reply src/main/java/org/somerville/swag/data/source/DBMapper.java:23]: Thanks Alan, I will look at researching this online

File: src/main/java/org/somerville/swag/data/entity/state/Guest.java (hunk 6 above)
Comment (sgoudham) [reply src/main/java/org/somerville/swag/data/entity/state/Guest.java:66]: Thank you!

**Review by erinmaria98 — APPROVED** [review]
Code looks a lot cleaner and has better readability 
The additions look great too.

**Pull request merged**